File: hooks/check-comments.mjs
#!/usr/bin/env node
// PreToolUse hook for Edit/Write/MultiEdit. Measures how much of the incoming
// text is comment and denies once when it crosses a threshold; an identical
// resubmit is accepted, on the basis that the model looked again and kept them.
//
// Runs on every file write, so: comment text is never logged (only sizes, paths,
// and decisions), and every path that is not a confident deny exits 0 silently.
import { existsSync, mkdirSync, readdirSync, statSync, unlinkSync, writeFileSync } from 'node:fs';
import { createHash } from 'node:crypto';
import { join } from 'node:path';
import { cfg, bool, log, rotateLog, readStdin, validSessionId, SESSIONS_DIR, SENTINEL_TTL_MS } from './common.mjs';
import { detectLang, isSkippedExt, extractComments, summarise } from './comments.mjs';
import { denyReason } from './guidance.mjs';

const DISABLED = cfg('CLAUDE_ELOQUENT_DISABLED', 'CLAUDE_PLUGIN_OPTION_DISABLED', '0');
const RATIO = parseFloat(cfg('CLAUDE_ELOQUENT_RATIO', 'CLAUDE_PLUGIN_OPTION_COMMENT_RATIO', '0.40'));
const MIN_CHARS = parseInt(cfg('CLAUDE_ELOQUENT_MIN_CHARS', 'CLAUDE_PLUGIN_OPTION_MIN_CHARS', '200'), 10);
const CHECK_BLOCK_LINES = cfg('CLAUDE_ELOQUENT_CHECK_BLOCK_LINES', 'CLAUDE_PLUGIN_OPTION_CHECK_BLOCK_LINES', '0');
const MAX_BLOCK_LINES = parseInt(cfg('CLAUDE_ELOQUENT_MAX_BLOCK_LINES', 'CLAUDE_PLUGIN_OPTION_MAX_BLOCK_LINES', '6'), 10);
const ALLOW_ON_RETRY = cfg('CLAUDE_ELOQUENT_ALLOW_ON_RETRY', 'CLAUDE_PLUGIN_OPTION_ALLOW_ON_RETRY', '1');
const EXTRA_SKIP_EXT = (process.env.CLAUDE_ELOQUENT_EXTRA_SKIP_EXT ?? '')
  .split(',').map(s => s.trim().replace(/^\./, '').toLowerCase()).filter(Boolean);

function analysedText(toolName, input) {
  if (toolName === 'Write') return typeof input.content === 'string' ? input.content : null;
  if (toolName === 'Edit') return typeof input.new_string === 'string' ? input.new_string : null;
  if (toolName === 'MultiEdit') {
    if (!Array.isArray(input.edits)) return null;
    return input.edits.map(e => (typeof e?.new_string === 'string' ? e.new_string : '')).join('\n');
  }
  return null;
}

// Sentinels expire so a long session cannot accumulate them, and so an edit
// revisited hours later is judged fresh rather than waved through.
function pruneSentinels(dir) {
  try {
    const now = Date.now();
    for (const name of readdirSync(dir)) {
      const full = join(dir, name);
      try {
        if (now - statSync(full).mtimeMs > SENTINEL_TTL_MS) unlinkSync(full);
      } catch { /* raced with another hook */ }
    }
  } catch { /* dir may not exist yet */ }
}

try {
  if (bool(DISABLED)) process.exit(0);

  const event = JSON.parse(readStdin());
  const toolName = event.tool_name ?? '';
  const input = event.tool_input ?? {};
  const filePath = input.file_path;

  if (!filePath || typeof filePath !== 'string') process.exit(0);

  if (event.cwd && existsSync(join(event.cwd, '.claude-eloquent-skip'))) {
    log('check', 'SKIP: disabled via .claude-eloquent-skip');
    process.exit(0);
  }

  if (isSkippedExt(filePath, EXTRA_SKIP_EXT)) {
    log('check', `SKIP: doc ext ${filePath}`);
    process.exit(0);
  }

  const lang = detectLang(filePath);
  if (!lang) process.exit(0);

  const text = analysedText(toolName, input);
  if (!text) process.exit(0);

  const result = extractComments(text, lang);
  const { ratio, longestBlock, blockCount } = summarise(result);

  const ratioTripped = result.totalChars >= MIN_CHARS && ratio > RATIO;
  const blockTripped = bool(CHECK_BLOCK_LINES) && longestBlock > MAX_BLOCK_LINES;
  if (!ratioTripped && !blockTripped) process.exit(0);

  rotateLog();

  const sessionId = event.session_id ?? '';
  const percent = Math.round(ratio * 100);
  const stats = `${filePath} ${percent}% of ${result.totalChars} chars, ${blockCount} blocks, longest ${longestBlock}`;

  // No usable session id means no sentinel, so deny-once would become
  // deny-always. Fail open instead.
  if (!validSessionId(sessionId)) {
    log('check', `SKIP: invalid session_id (would have denied: ${stats})`);
    process.exit(0);
  }

  const sessionDir = join(SESSIONS_DIR, sessionId);
  const key = createHash('sha256').update(`${filePath}\n${text}`).digest('hex').slice(0, 16);
  const sentinel = join(sessionDir, key);

  if (bool(ALLOW_ON_RETRY)) {
    try {
      mkdirSync(sessionDir, { recursive: true });
      pruneSentinels(sessionDir);
      // Left in place, so a third identical submit is allowed too.
      if (existsSync(sentinel)) {
        log('check', `ALLOW (retry): ${stats}`);
        process.exit(0);
      }
      writeFileSync(sentinel, '');
    } catch (err) {
      // Unwritable sentinel dir turns deny-once into deny-always, so allow.
      log('check', `ALLOW (no sentinel: ${err.message}): ${stats}`);
      process.exit(0);
    }
  }

  log('check', `DENY: ${stats}`);
  process.stdout.write(JSON.stringify({
    hookSpecificOutput: {
      hookEventName: 'PreToolUse',
      permissionDecision: 'deny',
      permissionDecisionReason: denyReason({ file: filePath, percent, blocks: blockCount, longestBlock, allowRetry: bool(ALLOW_ON_RETRY) }),
    },
  }));
  process.exit(0);
} catch (err) {
  log('check', `ERROR: unexpected failure: ${err.message}`);
  process.exit(0);
}

File: hooks/check_comments.py
#!/usr/bin/env python3
"""PreToolUse hook for Edit/Write/MultiEdit. Measures how much of the incoming
text is comment and denies once when it crosses a threshold; an identical
resubmit is accepted, on the basis that the model looked again and kept them.

Runs on every file write, so: comment text is never logged (only sizes, paths,
and decisions), and every path that is not a confident deny exits 0 silently.
"""

import hashlib
import json
import math
import os
import sys
import time

sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))

from common import (  # noqa: E402
    SENTINEL_TTL_S,
    SESSIONS_DIR,
    as_bool,
    cfg,
    log,
    read_stdin,
    valid_session_id,
)
from comments import detect_lang, extract_comments, is_skipped_ext, summarise  # noqa: E402
from guidance import deny_reason  # noqa: E402

DISABLED = cfg("CLAUDE_ELOQUENT_DISABLED", "CLAUDE_PLUGIN_OPTION_DISABLED", "0")
RATIO = float(cfg("CLAUDE_ELOQUENT_RATIO", "CLAUDE_PLUGIN_OPTION_COMMENT_RATIO", "0.40"))
MIN_CHARS = int(cfg("CLAUDE_ELOQUENT_MIN_CHARS", "CLAUDE_PLUGIN_OPTION_MIN_CHARS", "200"))
CHECK_BLOCK_LINES = cfg("CLAUDE_ELOQUENT_CHECK_BLOCK_LINES", "CLAUDE_PLUGIN_OPTION_CHECK_BLOCK_LINES", "0")
MAX_BLOCK_LINES = int(cfg("CLAUDE_ELOQUENT_MAX_BLOCK_LINES", "CLAUDE_PLUGIN_OPTION_MAX_BLOCK_LINES", "6"))
ALLOW_ON_RETRY = cfg("CLAUDE_ELOQUENT_ALLOW_ON_RETRY", "CLAUDE_PLUGIN_OPTION_ALLOW_ON_RETRY", "1")
EXTRA_SKIP_EXT = [
    s.strip().lstrip(".").lower()
    for s in (os.environ.get("CLAUDE_ELOQUENT_EXTRA_SKIP_EXT") or "").split(",")
    if s.strip()
]


def analysed_text(tool_name, tool_input):
    if tool_name == "Write":
        content = tool_input.get("content")
        return content if isinstance(content, str) else None
    if tool_name == "Edit":
        new_string = tool_input.get("new_string")
        return new_string if isinstance(new_string, str) else None
    if tool_name == "MultiEdit":
        edits = tool_input.get("edits")
        if not isinstance(edits, list):
            return None
        parts = []
        for edit in edits:
            new_string = edit.get("new_string") if isinstance(edit, dict) else None
            parts.append(new_string if isinstance(new_string, str) else "")
        return "\n".join(parts)
    return None


def prune_sentinels(directory):
    """Sentinels expire so a long session cannot accumulate them, and so an edit
    revisited hours later is judged fresh rather than waved through."""
    try:
        now = time.time()
        for name in os.listdir(directory):
            full = os.path.join(directory, name)
            try:
                if now - os.path.getmtime(full) > SENTINEL_TTL_S:
                    os.unlink(full)
            except OSError:
                pass  # raced with another hook
    except OSError:
        pass  # dir may not exist yet


try:
    if as_bool(DISABLED):
        sys.exit(0)

    payload, truncated = read_stdin()
    if truncated:
        log("check", "SKIP: payload over 8MB")
        sys.exit(0)

    event = json.loads(payload)
    tool_name = event.get("tool_name", "")
    tool_input = event.get("tool_input") or {}
    file_path = tool_input.get("file_path")

    if not file_path or not isinstance(file_path, str):
        sys.exit(0)

    cwd = event.get("cwd")
    if cwd and os.path.exists(os.path.join(cwd, ".claude-eloquent-skip")):
        sys.exit(0)

    if is_skipped_ext(file_path, EXTRA_SKIP_EXT):
        sys.exit(0)

    lang = detect_lang(file_path)
    if not lang:
        sys.exit(0)

    text = analysed_text(tool_name, tool_input)
    if not text:
        sys.exit(0)

    result = extract_comments(text, lang)
    stats_summary = summarise(result)
    ratio = stats_summary["ratio"]
    longest_block = stats_summary["longest_block"]
    block_count = stats_summary["block_count"]

    ratio_tripped = result["total_chars"] >= MIN_CHARS and ratio > RATIO
    block_tripped = as_bool(CHECK_BLOCK_LINES) and longest_block > MAX_BLOCK_LINES
    if not ratio_tripped and not block_tripped:
        sys.exit(0)

    session_id = event.get("session_id", "")
    # Half-up, matching the Node implementation this replaced; Python's round()
    # is banker's rounding and would report 70 where the old hook reported 71.
    percent = int(math.floor(ratio * 100 + 0.5))
    stats = "{} {}% of {} chars, {} blocks, longest {}".format(
        file_path, percent, result["total_chars"], block_count, longest_block
    )

    # No usable session id means no sentinel, so deny-once would become
    # deny-always. Fail open instead.
    if not valid_session_id(session_id):
        log("check", "SKIP: invalid session_id (would have denied: {})".format(stats))
        sys.exit(0)

    session_dir = os.path.join(SESSIONS_DIR, session_id)
    key = hashlib.sha256("{}\n{}".format(file_path, text).encode("utf-8")).hexdigest()[:16]
    sentinel = os.path.join(session_dir, key)

    if as_bool(ALLOW_ON_RETRY):
        try:
            os.makedirs(session_dir, exist_ok=True)
            prune_sentinels(session_dir)
            # Left in place, so a third identical submit is allowed too.
            if os.path.exists(sentinel):
                log("check", "ALLOW (retry): {}".format(stats))
                sys.exit(0)
            with open(sentinel, "w"):
                pass
        except SystemExit:
            raise
        except OSError as err:
            # Unwritable sentinel dir turns deny-once into deny-always, so allow.
            log("check", "ALLOW (no sentinel: {}): {}".format(err, stats))
            sys.exit(0)

    log("check", "DENY: {}".format(stats))
    sys.stdout.write(json.dumps({
        "hookSpecificOutput": {
            "hookEventName": "PreToolUse",
            "permissionDecision": "deny",
            "permissionDecisionReason": deny_reason(
                file=file_path,
                percent=percent,
                blocks=block_count,
                longest_block=longest_block,
                allow_retry=as_bool(ALLOW_ON_RETRY),
            ),
        },
    }, separators=(",", ":"), ensure_ascii=False))
    sys.exit(0)
except SystemExit:
    raise
except Exception as err:
    log("check", "ERROR: unexpected failure: {}".format(err))
    sys.exit(0)
